bindings/matrix-sdk-ffi/src/sliding_sync.rs

```diff
@@ -0,0 +1,580 @@
+            })
+            .collect())
+    }
+
+    pub fn sync(&self) -> Arc<StoppableSpawn> {
+        let inner = self.inner.clone();
+        let observer = self.observer.clone();
+        let spawn = Arc::new(StoppableSpawn::with_handle_ref(self.sync_handle.clone()));
+        let inner_spawn = spawn.clone();
+        {
+            let mut sync_handle = self.sync_handle.write().unwrap();
+
+            if let Some(handle) = sync_handle.take() {
+                handle.abort();
+            }
+
+            *sync_handle = Some(RUNTIME.spawn(async move {
+                let stream = inner.stream().await.expect("Doesn't fail.");
+                pin_mut!(stream);
+                loop {
+                    let update = match stream.next().await {
+                        Some(Ok(u)) => u,
+                        Some(Err(e)) => {
+                            // FIXME: send this over the FFI
+                            tracing::warn!("Sliding Sync failure: {:?}", e);
+                            continue;
+                        }
+                        None => {
+                            tracing::debug!("No update from loop, cancelled");
+                            break;
+                        }
+                    };
+                    if let Some(ref observer) = *observer.read().unwrap() {
+                        observer.did_receive_sync_update(update.into());
+                    } else {
+                        // when the observer has been removed
+                        // we cancel the loop
+                        inner_spawn.cancel();
+                        break;
+                    }
+                }
+            }));
+        }
+
+        spawn
+    }
+}
+
+#[derive(Clone)]
+pub struct SlidingSyncBuilder {
+    inner: MatrixSlidingSyncBuilder,
+    client: Client,
+}
+
+impl SlidingSyncBuilder {
+    pub fn homeserver(self: Arc<Self>, url: String) -> anyhow::Result<Arc<Self>> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.inner.homeserver(url.parse()?);
+        Ok(Arc::new(builder))
+    }
+
+    pub fn add_fullsync_view(self: Arc<Self>) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.inner.add_fullsync_view();
+        Arc::new(builder)
+    }
+
+    pub fn no_views(self: Arc<Self>) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        builder.inner.no_views();
+        Arc::new(builder)
+    }
+
+    pub fn add_view(self: Arc<Self>, v: Arc<SlidingSyncView>) -> Arc<Self> {
+        let mut builder = unwrap_or_clone_arc(self);
+        let view = unwrap_or_clone_arc(v);
+        builder.inner.add_view(view.inner);
+        Arc::new(builder)
+    }
+
+    pub fn build(self: Arc<Self>) -> anyhow::Result<Arc<SlidingSync>> {
+        let builder = unwrap_or_clone_arc(self);
+        Ok(Arc::new(SlidingSync::new(builder.inner.build()?, builder.client)))
+    }
```

Comment:
I'm uncomfortable with this `unwrap_or_clone_arc` + `Arc::new` pattern.

I feel like we could use `Arc::get_mut` at some point. Or maybe `Arc::make_mut`, which will mimic this behavior better.

Thoughts?

Reply:
Copied this pattern from @jplatte . Thoughts?

Reply:
@Hywan `Arc::make_mut` only gives you `&mut`, we need an owned value here. `unrap_or_clone_arc` exists in the standard library as `Arc::unwrap_or_clone`, it just hasn't stabilized yet (partially because the naming hasn't been discussed enough, it might for example be renamed to `make_owned`).

I think it's perfectly fine to have our own helper function for it, it is very simple.

Reply:
It's not the helper that annoys me, it's the unwrap or clone itself :-p. If `T` is costly to clone, running this quite often can have a performance impact. Shouldn't it better to avoid having the need for ownership here? Thoughts?